Alternative Pool Strategies

gregw opened the pull request:
Speculative proposal to make Pool have a pluggable strategy for #5217

gregw committed:
Speculative idea to make a pluggable Pool strategy

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@sbordet @lorban i don't hate it...

gregw committed:
Speculative idea to make a pluggable Pool strategy

 + javadoc

Signed-off-by: Greg Wilkins <gregw@webtide.com>

lorban commented:
I'm not going to comment on the implementation that is a bit drafty, but making the thread-local caching pluggable is surprisingly not very intrusive, so I don't mind it.

But I'm still not conviced by the usefulness of this change as I'm worried it seems like it will create a tendency to push down some logic that should be kept higher-up, like the randomness / round robinness that IMHO is better suited at a higher level.

Oh, and some unwise strategy has the potential of totally killing perf, like the LRU implementation.

gregw committed:
Speculative idea to make a pluggable Pool strategy

 + Added a ThreadLocalStrategy for a single cached item
 + Tell strategies about newly reserved entries
 + Fixed multiplexing test that was dependent on the impl of the cache

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@lorban yep the LRU impl really sux :)  Perhaps we shouldn't offer that one as the name is attractive, but hard to implement without garbage and/or contention.

Have a look again now, as I've created ThreadLocalStrategy for the common case of a cache of size 1.  So that will give us a performance improvement.

gregw committed:
Speculative idea to make a pluggable Pool strategy

 + added tests

Signed-off-by: Greg Wilkins <gregw@webtide.com>

lorban commented:
What I'm uncomfortable with in this change is the fact that the strategy can either augment the entry lookup, like the caching strategy does by adding a fastpath / slowpath behavior, or it can completely bypass the entry lookup... while still having the ability to fallback to the default lookup.

This sounds like a too coarse/permissive API that could easily be abused. I'm worried this could encourage turning the Pool class into a silver bullet in the long run, especially since I can't think of a use-case that can't be solved by wrapping it instead.

gregw commented:
@lorban But I think it is better to have Pool be a Shiny bullet (if not silver), rather than have the HttpClient connection pool code worrying about the details of how to do random/round-robin/etc.
The #5219 proposal is worse for me because it mixes up concerns: the client code should not need to worry about how the pool works.

Wrapping is a poor solution as it requires exposing public API of aquireAt(int) and you end up with the situation of unused cache mechanisms in the code that you hope the JIT will remove and that they wont confuse connection cache developers.

Finally, if your arguments hold, then that is a reason to remove the caching mechanism from Pool as that too could be done with a wrapper!   Strategy should be 0,1 or infinity.  The fact that we have thread local cache and round robin shows that we have at least 2 strategies... hence we should implement infinity.

gregw committed:
Feedback from review

 + Don't have a fallback iteration, instead make a SearchStrategy and DualStrategy

gregw commented:
@lorban I think I've fixed your reservation of having the fallback iterative search.
Instead that is now just a `SearchStrategy` and I have added a `DualStrategy` so we have

cacheSize ==0 results in `new SearchStrategy()`
cacheSize == 1 results in `new DualStrategy(new ThreadLocalStrategy(), new SearchStrategy())`
cacheSize > 1 results in `new DualStrategy(new ThreadLocalCacheStrategy(cacheSize), new SearchStrategy())`

a round robin pool is just `new RoundRobinStrategy()` as it does not need a search
random does, so it should also be a `new DualStrategy(new RandomStrategy(), new SearchStrategy())`

Naming could be worked on....
If you like we could have an `interface Strategy` and `interface CompleteStrategy extends Strategy` with the constructor needing a `CompleteStrategy`.
Search and RoundRobin are `CompleteStrategy`'s but the others are not.  Dual could be renamed `CachingStrategy`  where the second strategy has to be complete.   This would prevent bad combinations.

gregw committed:
Feedback from review

 + split strategies into Cache and Strategies

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -165,7 +201,8 @@ public Entry reserve(int allotment)
             Entry entry = new Entry();
-            sharedList.add(entry);
+            entries.add(entry);
+            strategy.released(entries, entry, true);
```
Why a `release()` inside `reserve()`?

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -559,4 +574,222 @@ public String toString()
+         * @param entry The entry to be release
+         * @param reusable true if the entry is reusable and will be put back in the pool.
+         */
+        default void released(List<Pool<T>.Entry> entries, Pool<T>.Entry entry, boolean reusable)
```
Just `release()` and implementations have to call `entry.tryRelease()` so that we're symmetric with `acquire()` above?

lorban replied:
I second @sbordet; and wouldn't the following signature be better?

 `void release(Pool<T>.Entry entry);`

without a list of entries (I can't see why it would be needed) and no default impl?

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -559,4 +574,222 @@ public String toString()
+        default void released(List<Pool<T>.Entry> entries, Pool<T>.Entry entry, boolean reusable)
+        {
+        }
+    }
```
Is not this class missing a `remove()` method?

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -559,4 +574,222 @@ public String toString()
+        }
+    }
+
+    public static class SearchStrategy<T> implements Strategy<T>
```
`LinearSearchStrategy`?

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -559,4 +574,222 @@ public String toString()
+
+    public static class RoundRobinCache<T> extends IndexedCached<T>
+    {
+        AtomicInteger index = new AtomicInteger();
```
`private final`?

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -559,4 +574,222 @@ public String toString()
+        }
+    }
+
+    public static class LeastRecentlyUsedStrategy<T> implements Strategy<T>
```
do we want to keep this strategy?

gregw replied:
I'll keep for now to test API during dev... but don't think we should include it in final merge... actually I might move it to a test to check that an external strategy can be applied without access to protected methods

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -559,4 +574,222 @@ public String toString()
+        }
+    }
+
+    public interface Strategy<T> extends Cache<T>
```
Are both `Strategy` and `Cache` needed?

gregw replied:
Strategy and Cache are needed if we want to make a distinction between impls that just try to find an entry and then give up (threadlocal, random) and those that do a search  ( linear, round-robin).

However, I too am not convinced the distinction is necessary.

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -176,14 +213,15 @@ public Entry reserve(int allotment)
      * @param idx the index of the entry to acquire.
      * @return the specified entry or null if there is none at the specified index or if it is not available.
      */
+    @Deprecated
```
If we go with this design, the RR connection pool should be rewritten to use the RR strategy and this method should be removed.

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -559,4 +574,222 @@ public String toString()
+        {
+            Objects.requireNonNull(cache);
+            Objects.requireNonNull(planB);
+            this.cache = cache;
```
`this.cache = Objects.requireNonNull(cache);`

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -559,4 +574,222 @@ public String toString()
+            Objects.requireNonNull(cache);
+            Objects.requireNonNull(planB);
+            this.cache = cache;
+            this.strategy = planB;
```
`this.strategy = Objects.requireNonNull(planB);`

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -197,32 +235,14 @@ public Entry acquireAt(int idx)
     /**
      * Acquire an entry from the pool.
      * Only enabled entries will be returned from this method and their enable method must not be called.
+     * The implementation first tries the pool strategy and then a brute force iteration over entries.
```
This comment isn't accurate anymore as the iteration is also part of the strategy.

lorban reviewed (changes requested):
I'm leaning on the path of being convinced by this `Strategy` approach, mostly because we could get rid of the `acquireAt()` method.

But the `Strategy` API needs some cleanup IMHO.

gregw committed:
Feedback from review

 + Added reserve and release

gregw commented:
@sbordet @lorban I've followed Simone's feedback and it is working... but I think it is really rather delicate and there are all sorts of ways that it can be used incorrectly.
I've put this back into draft as I think it has become too big and invasive and will be too difficult to use.

I'm not sure that making the strategies get involved with the actual mechanics of release/remove/reserve is a good thing.   Sure it makes the interface look symmetric (having `release` instead of `released`), but it is putting too much of the internals into the strategy.  You have to deal with if the first strategy releases, then the second doesn't know about it etc.  Multiplexing also makes composite release very complex.

Fundamentally I think it kind of goes against the atomic spirit of the pool, where the Entry is the primary atomic mechanism and when you act on it, you might find that somebody else has already acquired or closed it on you.  Thus I think a `released` notification on the strategy is much simpler and better than a `release` operation.   Note also that the `tryRelease` API need for the operation is not public.

So I am 99% sure that this is the right place to implement round robin / random style strategies.   I'm just 1% sure this is currently the right way to do it.

gregw committed:
Improved Pool Strategies:

+ reverted to post notifications for removed, reserved and released.
+ Added a few more strategies that need to be benchmarked, that use the list iterator.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@lorban @sbordet I know I shouldn't be playing with this... but....

I've updated to a simpler strategy that is notified after removed, reserved, released - as that makes combining strategies sensible.
However, I've then been experimenting with using `cowList.listIterator(int index)` for strategies that can have a random or round-robin starting point, but then iterator themselves if need be.    These might produce a bit more garbage, but there are also ways to reuse the iterators.   Using such strategies means that we don't need to combine with a LinearSearch, so perhaps we can go back to strategies that actually do the operation on the list.

Anyway, this is still a draft, but have a look if you are interested.... but only once jetty-10 is ready to release.

cheers

gregw committed:
Testing all the different strategies

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
More simplifications and made LRU work (ish)

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
javadoc

gregw committed:
More javadoc

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
JMH Test

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
one strategy

Signed-off-by: gregw <gregw@webtide.com>

gregw committed:
test

Signed-off-by: gregw <gregw@webtide.com>

gregw committed:
Split implementations:

 + pluggable strategies
 + hard coded

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
More benchmarks

gregw commented:
Some strange benchmark results!

 + I narrowed testing down to 4 key strategies: Linear, Random, ThreadLocal, RoundRobin
 + I wrote a single strategy that could be configured to do any of those: OneStrategyToRuleThemAll.   It works fine (although it's round robin impl has a hack that wont work in general).
 + So I thought what if I just hard coded the OSTRTA into a Pool class with some inheritance to make the different implementations!  The results are horrid:

```text
           (POOL_TYPE)           Ops/s          Error   Units     alloc rate
                linear    12388037.627 ±  4447723.249   ops/s     0.001 ±   0.005  MB/sec
        OSTRTA(LINEAR)    11712109.534 ± 18050206.735   ops/s     0.001 ±   0.008  MB/sec
           Pool.Linear     9076780.369 ±   772761.089   ops/s     0.001 ±   0.009  MB/sec
    threadlocal+linear    19955813.938 ± 15912431.612   ops/s     0.001 ±   0.008  MB/sec
  OSTRTA(THREAD_LOCAL)    18986205.894 ± 34689639.507   ops/s     0.001 ±   0.008  MB/sec
           Pool.Thread    10881925.059 ±  6078295.882   ops/s     0.001 ±   0.008  MB/sec
      random-iteration    17877890.452 ±  2485249.042   ops/s   377.917 ± 403.364  MB/sec
        OSTRTA(RANDOM)    16392327.062 ±  9050945.178   ops/s     0.001 ±   0.009  MB/sec
           Pool.Random     9314759.324 ± 16094548.697   ops/s     0.001 ±   0.008  MB/sec
     roundrobin+linear     8846940.752 ±  9466144.785   ops/s   124.635 ± 153.511  MB/sec
   OSTRTA(ROUND_ROBIN)     8709583.310 ±  4613331.332   ops/s   122.741 ± 110.409  MB/sec
       Pool.RoundRobin     8079087.371 ± 17119090.317   ops/s     0.001 ±   0.008  MB/sec
```

Ignore the garbage generated by round robin, that's a lambda I can get rid of.

All the inherited behaviour ones:  Pool.Linear, Pool.Thread etc.  are the slowest!  Virtual calls I guess.


So I can see 2 ways forward:
1. Go back to the pluggable strategy pool and just provide the 4 key strategies (will replace random-iteration with one that doesn't allocate)
2. Go to a single Pool class with a passed in Mode (like OSTRTA) 

I'm guessing @sbordet likes 1. and @lorban likes 2. ??? I'm kind of on the fence and need to be nudged either way.

gregw commented:
So the results show us the way:
```
Benchmark                                                        (POOL_TYPE)  (SIZE)   Mode  Cnt         Score          Error  Units
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy              linear      16  thrpt    3  16627512.074 ± 17642452.179  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy         Pool.Linear      16  thrpt    3  18129265.620 ±  4463027.935  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy  threadlocal+linear      16  thrpt    3  31143782.885 ± 42795850.366  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy         Pool.Thread      16  thrpt    3  43294021.851 ± 60671099.356  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy       random+linear      16  thrpt    3  20697605.670 ±  1380766.015  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy         Pool.Random      16  thrpt    3  24497173.155 ±  9343959.400  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy   roundrobin+linear      16  thrpt    3   7313809.523 ±   172708.836  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy     Pool.RoundRobin      16  thrpt    3   7213369.700 ±   214676.084  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy     threadid+linear      16  thrpt    3  48749994.419 ± 76484935.329  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy       Pool.ThreadId      16  thrpt    3  39440370.903 ± 72046403.132  ops/s
```

The hard coded strategies match the pluggable strategies, except for threadlocal, in which case it is a lot better.    So given the simplicity of
the built in approach, I will prepare the PR assuming that.

gregw committed:
Built in strategy

gregw committed:
removed strategies version and simplified to single configurable solution.

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw commented:
@sbordet I think this is ready to use in the connection pools, but it needs changes to their constructors, so I might let you do it?

gregw commented:
Final benchmark results:
```
Benchmark                                                 (CACHE)      (POOL_TYPE)  (SIZE)   Mode  Cnt         Score           Error  Units
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy    false      Pool.Linear       4  thrpt    3  19726151.258 ±   4667777.948  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy    false      Pool.Linear      16  thrpt    3  17818008.006 ±   9139880.033  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy    false      Pool.Random       4  thrpt    3  17996978.870 ±   3800769.899  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy    false      Pool.Random      16  thrpt    3  26759187.237 ±   3575103.399  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy    false  Pool.RoundRobin       4  thrpt    3   8484006.729 ±    890613.362  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy    false  Pool.RoundRobin      16  thrpt    3   8528120.962 ±   3044297.564  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy    false    Pool.ThreadId       4  thrpt    3  17647364.947 ±   2405801.834  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy    false    Pool.ThreadId      16  thrpt    3  45474461.806 ± 115516916.830  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy     true      Pool.Linear       4  thrpt    3  19061961.075 ±   9416326.911  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy     true      Pool.Linear      16  thrpt    3  40489787.862 ±  19745433.024  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy     true      Pool.Random       4  thrpt    3  18678870.374 ±  20741555.087  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy     true      Pool.Random      16  thrpt    3  52719122.996 ±  97184200.471  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy     true  Pool.RoundRobin       4  thrpt    3  17302566.141 ±    746359.957  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy     true  Pool.RoundRobin      16  thrpt    3  45292108.520 ±  37436891.524  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy     true    Pool.ThreadId       4  thrpt    3  21522511.381 ±   3130277.130  ops/s
PoolStrategyBenchmark.testAcquireReleasePoolWithStrategy     true    Pool.ThreadId      16  thrpt    3  39812967.055 ±  81267475.628  ops/s
```

gregw commented:
![Screenshot from 2020-09-10 16-05-59](https://user-images.githubusercontent.com/70829/92742086-938b5680-f37f-11ea-9dfc-85594994ae07.png)

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -65,29 +75,74 @@
      * When an entry can't be found in the cache, the global list is iterated
      * normally so the cache has no visible effect besides performance.
      */
```
You should move this comment around since you also moved the ThreadLocal variable it speaks about.

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -176,14 +231,15 @@ public Entry reserve(int allotment)
      * @return the specified entry or null if there is none at the specified index or if it is not available.
      */
+    @Deprecated
     public Entry acquireAt(int idx)
```
This method should go and any caller should be replaced with a Pool using the RR strategy.

gregw replied:
Yep - waiting for @sbordet to help rework connection pool.

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -204,25 +260,51 @@ public Entry acquire()
+            case THREAD_ID:
+                return (int)(Thread.currentThread().getId() % size);
+            default:
+                throw new IllegalArgumentException();
```
I'd add a message to the exception, something like
```
throw new IllegalArgumentException("Invalid strategy : " + strategy);
```

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -500,6 +575,13 @@ boolean tryRelease()
             return !(overUsed && newMultiplexingCount == 0);
         }
 
+        public boolean isOverUsed()
```
This doesn't seem to be used.

gregw replied:
I did have a use for it.... but can't remember now... was probably some esoteric strategy or perhaps a unit test.
I still think it is a nice to have along with the `isIdle` type methods.

sbordet replied:
+1 for keeping the method for JMX and such.

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -452,7 +527,7 @@ public boolean remove()
          * false otherwise.
          */
-        boolean tryAcquire()
+        public boolean tryAcquire()
```
This doesn't have to be `public`.

gregw replied:
Ah yes! It needed to be public for strategies... but not now!

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -345,45 +417,48 @@ public void close()
         // hi: positive=open/maxUsage counter; negative=closed; MIN_VALUE pending
         // lo: multiplexing counter
         private final AtomicBiInteger state;
+        private final int index;
```
This index doesn't seem to be used.

gregw replied:
oops yes - that was a broken hack for another esoteric round robin

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -549,12 +631,17 @@ public int getUsageCount()
             long encoded = state.get();
-            return String.format("%s@%x{usage=%d/%d,multiplex=%d/%d,pooled=%s}",
+            int hi = AtomicBiInteger.getHi(encoded);
+            int lo = AtomicBiInteger.getLo(encoded);
```
I would rename these variables to the actual name of their value: `usageCount` and `multiplexingCount`.

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -65,29 +75,74 @@
+         * The Linear strategy looks for an entry always starting from the first entry.
+         * It will favour the early entries in the pool, but may contend on them more.
+         */
+        LINEAR,
```
Since all the other names of this enum describe the starting index, I'd rename this one to something like `INDEX_ZERO`.

gregw replied:
How about `FIRST`?

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -65,29 +75,74 @@
     private final AtomicInteger pending = new AtomicInteger();
+    private final Strategy strategy;
+    private final ThreadLocal<Entry> cache;
+    private final AtomicInteger next;
```
`AtomicInteger nextIndex`?

lorban commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -65,29 +75,74 @@
-        else
-            this.cache = null;
+        this.strategy = strategy;
+        this.cache = cache ? new ThreadLocal() : null;
```
`new ThreadLocal<>()`?

gregw replied:
Don't need the threadlocal if not caching

lorban reviewed (changes requested):
Besides a few small cleanups, I like that the all-contained-and-compact design was kept. But the PR's name is now wrong.

 LGTM.

gregw committed:
updates from review

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
better javadoc

Signed-off-by: Greg Wilkins <gregw@webtide.com>

gregw committed:
Merge branch 'jetty-9.4.x' into jetty-9.4.x-PoolStrategy

gregw committed:
Updated ConnectionPool classes to use Pool strategies

sbordet commented on VERSION.txt:
```diff
@@ -1198,7 +1198,7 @@ jetty-9.4.4.v20170414 - 14 April 2017
  + 1436 NullPointerException when calling changeSessionId
  + 1439 Allow UNC paths to function as Resource bases
- + 1440 Improve lock contention for low resources scheduling strategy
+ + 1440 Improve lock contention for low resources scheduling strategyType
```
Revert this, it was a search/replace gone wrong.

gregw replied:
Ah ooops.   That's an IDE being too smart!

sbordet commented on jetty-documentation/src/main/asciidoc/administration/logging/jetty-server-dump.adoc:
```diff
@@ -105,10 +105,10 @@ Server@59906517{STARTED}[9.4.32-SNAPSHOT] - STARTED
+|  |  |  +> app//org.eclipse.jetty.util.thread.strategyType.EatWhatYouKill.runTask(EatWhatYouKill.java:336)
+|  |  |  +> app//org.eclipse.jetty.util.thread.strategyType.EatWhatYouKill.doProduce(EatWhatYouKill.java:313)
+|  |  |  +> app//org.eclipse.jetty.util.thread.strategyType.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171)
+|  |  |  +> app//org.eclipse.jetty.util.thread.strategyType.EatWhatYouKill.run(EatWhatYouKill.java:129)
```
Revert, bad search/replace.

sbordet commented on jetty-documentation/src/main/asciidoc/configuring/jsp/configuring-jsp.adoc:
```diff
@@ -163,7 +163,7 @@ Set all parameters on the `org.apache.jasper.servlet.JspServlet` instance define
 ____
 [NOTE]
-Be careful: for all of these parameters, if the value you set doesn't take effect, try using all lower case instead of camel case, or capitalizing only some of the words in the name, as JSP is inconsistent in its parameter naming strategy.
+Be careful: for all of these parameters, if the value you set doesn't take effect, try using all lower case instead of camel case, or capitalizing only some of the words in the name, as JSP is inconsistent in its parameter naming strategyType.
```
Revert, bad search/replace.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -56,38 +51,98 @@
+                return listIterator();
+            }
+        }
+    };
```
Overriding `listIterator(int)` is not necessary anymore, since we never use it, and we don't leak the `entries` field.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -56,38 +51,98 @@
      * undersized compared to the load applied on it.
      * When an entry can't be found in the cache, the global list is iterated
-     * normally so the cache has no visible effect besides performance.
+     * with the normal strategy so the cache has no visible effect besides performance.
```
There is no "normal" strategy, I would rephrase: "is iterated with the configured StrategyType".

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -56,38 +51,98 @@
     private volatile int maxMultiplex = 1;
     private volatile int maxUsageCount = -1;
 
+    public enum StrategyType
```
Javadocs.

sbordet commented on jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:
```diff
@@ -176,14 +231,15 @@ public Entry reserve(int allotment)
      * @param idx the index of the entry to acquire.
      * @return the specified entry or null if there is none at the specified index or if it is not available.
      */
+    @Deprecated
```
Add `@deprecated` javadoc too.

gregw replied:
I guess we can't just remove it... will do in 10

sbordet commented on jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java:
```diff
@@ -40,20 +50,27 @@
+        data.add(new Object[]{(Factory)s -> new Pool<String>(FIRST, s)});
+        data.add(new Object[]{(Factory)s -> new Pool<String>(RANDOM, s)});
+        data.add(new Object[]{(Factory)s -> new Pool<String>(FIRST, s, true)});
+        data.add(new Object[]{(Factory)s -> new Pool<String>(ROUND_ROBIN, s)});
```
new Pool<>(...) is enough, remove type parameter.

sbordet commented on jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java:
```diff
@@ -331,34 +347,44 @@ public void testMaxUsageCount(int cacheSize)
-        Pool<String> pool = new Pool<>(1, cacheSize);
+        Pool<String> pool = factory.getPool(1);
         pool.setMaxMultiplex(2);
         pool.reserve(-1).enable("aaa", false);
```
This test contains an unused `e2` variable that I would instead add to the asserts, for example, to confirm that it is in use even after removing `e1`, or that `e2 == e1`, etc.

sbordet commented on jetty-util/src/test/java/org/eclipse/jetty/util/PoolTest.java:
```diff
@@ -506,10 +532,10 @@ public void testMultiplexMaxUsageReachedAcquireThenReleaseThenRemove(int cacheSi
+        Pool<String> pool = factory.getPool(1);
         pool.setMaxMultiplex(2);
         pool.setMaxUsageCount(10);
         pool.reserve(-1).enable("aaa", false);
```
This test contains an unused variable `e2` that I would use in asserts, at least to verify that `e2 == e1`.

sbordet committed:
Small javadocs fixes.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw committed:
Merge branch 'jetty-9.4.x' into jetty-9.4.x-PoolStrategy

gregw committed:
Updates from review

sbordet reviewed (approved):
I'd like a minimal javadocs for `Pool.StrategyType`. Other than that, LGTM.

gregw committed:
javadoc
